### PR TITLE
Refactoring sabot and making it extensible

### PR DESCRIPTION
**Pull request by rmannibucau**
- Reorganising the project to make more explicit what is internals or not
- Adding Decrypter API
- Better alternative support
- Configuration metamodel available in the extension (so injectable)
- Merging TomEE integration to avoid to need another module and making it part of the SPI of the project

### CODE REVIEW

**Comment by jeanouii** [comment]
@rmannibucau I was looking at the PR and have a couple of comments/asks

Very well done with the Cipher and the improvements. Very useful.
- can you separate in more than one big PR?
- mixing refactoring and features/fixes makes it almost impossible to review, merge or reject a sub part

Also regarding this comment

> Merging TomEE integration to avoid to need another module and making it part of the SPI of the project

The goal of the previous release and changes was to explicitly split the TomEE specific code into a separate module so that Sabot can be used with another application server.

This makes total sense for an open source library and needs to remain.
If the listener or current way to plugin the TomEE extension isn't optimal, feel free to propose a better approach.

Thanks

**Comment by rmannibucau** [comment]
TomEE is fully optional with this PR and sabot stay compatible with wildfly but it is working OOTB if running in TomEE in a reliable way.

File: src/main/java/org/tomitribe/sabot/internal/integration/TomEE.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.tomitribe.sabot.internal.integration;
+
+import org.apache.openejb.loader.SystemInstance;
```
Comment (jeanouii) [inline src/main/java/org/tomitribe/sabot/internal/integration/TomEE.java:19]: Another import here

**Comment by jeanouii** [comment]
The problem with having everything in the same module is that with the dependencies available in the core, it will likely be updated some day to add a required link to TomEE which we don't want.

It's probably more convenient from a dependency point of view to have everything in the same module but it's not as clean in my opinion, even if with CDI tricks you can make sure TomEE specific classes aren't loaded when not in TomEE.

**Comment by rmannibucau** [comment]
up to you

**Comment by jeanouii** [comment]
It's not up to me lol
It's a community project so I'm happy to do as the community wants.

**Comment by AndyGee** [comment]
Pulled the patch and testing on project. What happened to the modules?

**Comment by rmannibucau** [comment]
No need of modules for the light integration we need IMO.

**Comment by AndyGee** [comment]
The idea was to add more modules for other servers as examples, and keep app server tests completely out of sabot

**Comment by rmannibucau** [comment]
I got it, don't share the choice but no problem to not merge this PR.

**Comment by AndyGee** [comment]
The rest of it looks really good, but I'd like to keep modules and not reference anything app server specific in sabot pom

**Comment by AndyGee** [comment]
I guess it should be usable in any CDI enabled app

**Comment by AndyGee** [comment]
Good case for another module maybe ;-)

**Comment by tomitribe-dev** [comment]
Can one of the admins verify this patch?

**Comment by rmannibucau** [comment]
outdated

**Pull request closed without merging**